File: system/ulib/mx/socket.cc
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <mx/socket.h>

#include <magenta/syscalls.h>

namespace mx {

mx_status_t socket::create(uint32_t flags, socket* endpoint0, socket* endpoint1) {
    mx_handle_t h0 = MX_HANDLE_INVALID, h1 = MX_HANDLE_INVALID;
    mx_status_t result = mx_socket_create(flags, &h0, &h1);
    endpoint0->reset(h0);
    endpoint1->reset(h1);
    return result;
}

} // namespace mx

File: src/connectivity/overnet/lib/core/src/fidl_tests/socket.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#![cfg(test)]

use super::{Fixture, Target};
use crate::test_util::NodeIdGenerator;
use async_trait::async_trait;
use fidl_handle_tests::{socket, LoggingFixture};

struct SockFixture {
    fixture: Fixture,
}

impl SockFixture {
    async fn new(node_id_gen: NodeIdGenerator) -> SockFixture {
        SockFixture { fixture: Fixture::new(node_id_gen).await }
    }
}

#[async_trait]
impl socket::Fixture for SockFixture {
    async fn create_handles(&self, opts: fidl::SocketOpts) -> (fidl::Socket, fidl::Socket) {
        let (local, remote) = fidl::Socket::create(opts).unwrap();
        (local, self.fixture.distribute_handle(remote, Target::B).await)
    }
}

impl LoggingFixture for SockFixture {
    fn log(&mut self, msg: &str) {
        self.fixture.log(msg)
    }
}

#[fuchsia_async::run(1, test)]
async fn fidl_socket_tests(run: usize) {
    crate::test_util::init();
    let node_id_gen = NodeIdGenerator::new("fidl_socket_tests", run);
    let fixture = SockFixture::new(node_id_gen).await;
    socket::run(fixture).await
}
